File: portfolio-management-system/src/models/transaction.py
"""
Transaction Data Model

Migrated from COBOL copybook: src/copybook/common/TRNREC.cpy

Original COBOL structure:
- TRN-KEY: Composite key (DATE + TIME + PORTFOLIO-ID + SEQUENCE-NO)
- TRN-DATA: Transaction details (investment, type, quantity, price, amount, currency, status)
- TRN-AUDIT: Audit trail (process date, user)

COBOL Data Types Mapping:
- PIC X(n) -> str
- PIC S9(11)V9(4) COMP-3 -> Decimal (packed decimal)
- PIC S9(13)V9(2) COMP-3 -> Decimal (packed decimal)
"""

from datetime import datetime, date, time
from decimal import Decimal
from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field, field_validator
from sqlalchemy import Column, String, Date, Time, Numeric, DateTime, Index

from src.models.base import Base


class TransactionType(str, Enum):
    """
    Transaction type codes.
    
    Migrated from COBOL 88-level conditions:
    - TRN-TYPE-BUY   VALUE 'BU'
    - TRN-TYPE-SELL  VALUE 'SL'
    - TRN-TYPE-TRANS VALUE 'TR'
    - TRN-TYPE-FEE   VALUE 'FE'
    """
    BUY = "BU"
    SELL = "SL"
    TRANSFER = "TR"
    FEE = "FE"


class TransactionStatus(str, Enum):
    """
    Transaction status codes.
    
    Migrated from COBOL 88-level conditions:
    - TRN-STATUS-PEND VALUE 'P'
    - TRN-STATUS-DONE VALUE 'D'
    - TRN-STATUS-FAIL VALUE 'F'
    - TRN-STATUS-REV  VALUE 'R'
    """
    PENDING = "P"
    DONE = "D"
    FAILED = "F"
    REVERSED = "R"


class TransactionRecord(BaseModel):
    """
    Pydantic model for transaction record validation.
    
    Preserves all field definitions from TRNREC.cpy with Python type mappings.
    """
    
    # TRN-KEY fields
    trn_date: date = Field(
        ...,
        description="Transaction date (YYYYMMDD format in COBOL)"
    )
    trn_time: time = Field(
        ...,
        description="Transaction time (HHMMSS format in COBOL)"
    )
    trn_portfolio_id: str = Field(
        ...,
        max_length=8,
        description="Portfolio identifier"
    )
    trn_sequence_no: str = Field(
        ...,
        max_length=6,
        description="Sequence number for multiple transactions"
    )
    
    # TRN-DATA fields
    trn_investment_id: str = Field(
        ...,
        max_length=10,
        description="Investment identifier"
    )
    trn_type: TransactionType = Field(
        ...,
        description="Transaction type: BU=Buy, SL=Sell, TR=Transfer, FE=Fee"
    )
    trn_quantity: Decimal = Field(
        ...,
        max_digits=15,
        decimal_places=4,
        description="Transaction quantity (PIC S9(11)V9(4) COMP-3)"
    )
    trn_price: Decimal = Field(
        ...,
        max_digits=15,
        decimal_places=4,
        description="Transaction price (PIC S9(11)V9(4) COMP-3)"
    )
    trn_amount: Decimal = Field(
        ...,
        max_digits=15,
        decimal_places=2,
        description="Transaction amount (PIC S9(13)V9(2) COMP-3)"
    )
    trn_currency: str = Field(
        ...,
        max_length=3,
        description="Currency code (ISO 4217)"
    )
    trn_status: TransactionStatus = Field(
        ...,
        description="Transaction status: P=Pending, D=Done, F=Failed, R=Reversed"
    )
    
    # TRN-AUDIT fields
    trn_process_date: Optional[datetime] = Field(
        None,
        description="Processing timestamp"
    )
    trn_process_user: Optional[str] = Field(
        None,
        max_length=8,
        description="Processing user ID"
    )

    @field_validator("trn_portfolio_id", "trn_sequence_no", "trn_investment_id", "trn_currency")
    @classmethod
    def strip_and_upper(cls, v: str) -> str:
        """Strip whitespace and convert to uppercase for consistency with COBOL."""
        if v:
            return v.strip().upper()
        return v

    @field_validator("trn_quantity", "trn_price", "trn_amount", mode="before")
    @classmethod
    def convert_to_decimal(cls, v) -> Decimal:
        """Convert numeric values to Decimal for precision."""
        if v is None:
            return Decimal("0")
        return Decimal(str(v))

    class Config:
        """Pydantic configuration."""
        json_encoders = {
            Decimal: str,
            date: lambda v: v.strftime("%Y%m%d"),
            time: lambda v: v.strftime("%H%M%S"),
            datetime: lambda v: v.isoformat(),
        }


class Transaction(Base):
    """
    SQLAlchemy ORM model for transaction records.
    
    Maps to PostgreSQL table: transactions
    Replaces VSAM TRANHIST file and DB2 TRANSACTION_HISTORY table.
    """
    __tablename__ = "transactions"

    # Primary key fields (composite key from TRN-KEY)
    trn_date = Column(Date, primary_key=True, nullable=False)
    trn_time = Column(Time, primary_key=True, nullable=False)
    trn_portfolio_id = Column(String(8), primary_key=True, nullable=False)
    trn_sequence_no = Column(String(6), primary_key=True, nullable=False)

    # Transaction data fields
    trn_investment_id = Column(String(10), nullable=False)
    trn_type = Column(String(2), nullable=False)
    trn_quantity = Column(Numeric(15, 4), nullable=False)
    trn_price = Column(Numeric(15, 4), nullable=False)
    trn_amount = Column(Numeric(15, 2), nullable=False)
    trn_currency = Column(String(3), nullable=False, default="USD")
    trn_status = Column(String(1), nullable=False, default="P")

    # Audit fields
    trn_process_date = Column(DateTime, nullable=True)
    trn_process_user = Column(String(8), nullable=True)

    # Indexes for common access patterns
    __table_args__ = (
        Index("idx_trn_portfolio_date", "trn_portfolio_id", "trn_date"),
        Index("idx_trn_investment", "trn_investment_id", "trn_date"),
        Index("idx_trn_status", "trn_status", "trn_date"),
    )

    def __repr__(self) -> str:
        return (
            f"<Transaction(portfolio={self.trn_portfolio_id}, "
            f"date={self.trn_date}, type={self.trn_type}, "
            f"amount={self.trn_amount})>"
        )

    def to_pydantic(self) -> TransactionRecord:
        """Convert SQLAlchemy model to Pydantic model for validation/serialization."""
        return TransactionRecord(
            trn_date=self.trn_date,
            trn_time=self.trn_time,
            trn_portfolio_id=self.trn_portfolio_id,
            trn_sequence_no=self.trn_sequence_no,
            trn_investment_id=self.trn_investment_id,
            trn_type=TransactionType(self.trn_type),
            trn_quantity=self.trn_quantity,
            trn_price=self.trn_price,
            trn_amount=self.trn_amount,
            trn_currency=self.trn_currency,
            trn_status=TransactionStatus(self.trn_status),
            trn_process_date=self.trn_process_date,
            trn_process_user=self.trn_process_user,
        )

    @classmethod
    def from_pydantic(cls, record: TransactionRecord) -> "Transaction":
        """Create SQLAlchemy model from Pydantic model."""
        return cls(
            trn_date=record.trn_date,
            trn_time=record.trn_time,
            trn_portfolio_id=record.trn_portfolio_id,
            trn_sequence_no=record.trn_sequence_no,
            trn_investment_id=record.trn_investment_id,
            trn_type=record.trn_type.value,
            trn_quantity=record.trn_quantity,
            trn_price=record.trn_price,
            trn_amount=record.trn_amount,
            trn_currency=record.trn_currency,
            trn_status=record.trn_status.value,
            trn_process_date=record.trn_process_date,
            trn_process_user=record.trn_process_user,
        )

File: modernized-investment-portfolio/src/main/java/com/investment/portfolio/entity/Transaction.java
package com.investment.portfolio.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

/**
 * JPA Entity mapping for the Transaction History table.
 *
 * COBOL Source: TRNREC.cpy (TRANSACTION-RECORD)
 *   TRN-KEY: TRN-DATE + TRN-TIME + TRN-PORTFOLIO-ID + TRN-SEQUENCE-NO
 *   TRN-DATA: TRN-INVESTMENT-ID, TRN-TYPE, TRN-QUANTITY, TRN-PRICE,
 *             TRN-AMOUNT, TRN-CURRENCY, TRN-STATUS
 *   TRN-AUDIT: TRN-PROCESS-DATE, TRN-PROCESS-USER
 *
 * DB2 Source: db2-definitions.sql (TRANSACTION_HISTORY)
 *   PK: TRANSACTION_ID
 *   FK: PORTFOLIO_ID -> PORTFOLIO_MASTER(PORTFOLIO_ID)
 *
 * Transaction types: BU=Buy, SL=Sell, TR=Transfer, FE=Fee
 * Status codes: P=Processed, F=Failed, R=Reversed
 * Transaction ID format: YYYYMMDDHHMMSS + 6-digit sequence
 */
@Entity
@Table(name = "transaction_history")
public class Transaction {

    /**
     * Transaction Identifier.
     * COBOL: TRN-DATE + TRN-TIME + TRN-PORTFOLIO-ID + TRN-SEQUENCE-NO
     * DB2: TRANSACTION_ID CHAR(20) NOT NULL
     * Format: YYYYMMDDHHMMSS + 6-digit sequence
     */
    @Id
    @Column(name = "transaction_id", length = 20, nullable = false)
    @NotNull
    @Size(max = 20)
    private String transactionId;

    /**
     * Portfolio Identifier (FK to PORTFOLIO_MASTER).
     * COBOL: TRN-PORTFOLIO-ID PIC X(08)
     * DB2: PORTFOLIO_ID CHAR(8) NOT NULL
     */
    @Column(name = "portfolio_id", length = 8, nullable = false, insertable = false, updatable = false)
    private String portfolioId;

    /**
     * Transaction Date.
     * COBOL: TRN-DATE PIC X(08) (YYYYMMDD)
     * DB2: TRANSACTION_DATE DATE NOT NULL
     */
    @Column(name = "transaction_date", nullable = false)
    @NotNull
    private LocalDate transactionDate;

    /**
     * Transaction Time.
     * COBOL: TRN-TIME PIC X(06) (HHMMSS)
     * DB2: TRANSACTION_TIME TIME NOT NULL
     */
    @Column(name = "transaction_time", nullable = false)
    @NotNull
    private LocalTime transactionTime;

    /**
     * Investment Identifier.
     * COBOL: TRN-INVESTMENT-ID PIC X(10)
     * DB2: INVESTMENT_ID CHAR(10) NOT NULL
     */
    @Column(name = "investment_id", length = 10, nullable = false)
    @NotNull
    @Size(max = 10)
    private String investmentId;

    /**
     * Transaction Type.
     * COBOL: TRN-TYPE PIC X(02) — BU=Buy, SL=Sell, TR=Transfer, FE=Fee
     * DB2: TRANSACTION_TYPE CHAR(2) NOT NULL
     */
    @Column(name = "transaction_type", length = 2, nullable = false)
    @NotNull
    @Size(max = 2)
    private String transactionType;

    /**
     * Transaction Quantity.
     * COBOL: TRN-QUANTITY PIC S9(11)V9(4) COMP-3
     * DB2: QUANTITY DECIMAL(18,4) NOT NULL
     */
    @Column(name = "quantity", precision = 18, scale = 4, nullable = false)
    @NotNull
    private BigDecimal quantity;

    /**
     * Transaction Price per unit.
     * COBOL: TRN-PRICE PIC S9(11)V9(4) COMP-3
     * DB2: PRICE DECIMAL(18,4) NOT NULL
     */
    @Column(name = "price", precision = 18, scale = 4, nullable = false)
    @NotNull
    private BigDecimal price;

    /**
     * Transaction Amount.
     * COBOL: TRN-AMOUNT PIC S9(13)V9(2) COMP-3
     * DB2: AMOUNT DECIMAL(18,2) NOT NULL
     */
    @Column(name = "amount", precision = 18, scale = 2, nullable = false)
    @NotNull
    private BigDecimal amount;

    /**
     * Currency Code (e.g. USD, EUR, GBP).
     * COBOL: TRN-CURRENCY PIC X(03)
     * DB2: CURRENCY_CODE CHAR(3) NOT NULL
     */
    @Column(name = "currency_code", length = 3, nullable = false)
    @NotNull
    @Size(max = 3)
    private String currencyCode;

    /**
     * Transaction Status.
     * COBOL: TRN-STATUS PIC X(01) — P=Pending/Processed, F=Failed, R=Reversed
     * DB2: STATUS CHAR(1) NOT NULL
     */
    @Column(name = "status", length = 1, nullable = false)
    @NotNull
    @Size(max = 1)
    private String status;

    /**
     * Processing Timestamp (audit field).
     * COBOL: TRN-PROCESS-DATE PIC X(26)
     * DB2: PROCESS_DATE TIMESTAMP NOT NULL
     */
    @Column(name = "process_date", nullable = false)
    @NotNull
    private LocalDateTime processDate;

    /**
     * Processing User (audit field).
     * COBOL: TRN-PROCESS-USER PIC X(08)
     * DB2: PROCESS_USER VARCHAR(8) NOT NULL
     */
    @Column(name = "process_user", length = 8, nullable = false)
    @NotNull
    @Size(max = 8)
    private String processUser;

    /**
     * Many-to-one relationship to Portfolio Master.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "portfolio_id", nullable = false)
    private PortfolioMaster portfolio;

    public Transaction() {
    }

    // --- Getters and Setters ---

    public String getTransactionId() {
        return transactionId;
    }

    public void setTransactionId(String transactionId) {
        this.transactionId = transactionId;
    }

    public String getPortfolioId() {
        return portfolioId;
    }

    public LocalDate getTransactionDate() {
        return transactionDate;
    }

    public void setTransactionDate(LocalDate transactionDate) {
        this.transactionDate = transactionDate;
    }

    public LocalTime getTransactionTime() {
        return transactionTime;
    }

    public void setTransactionTime(LocalTime transactionTime) {
        this.transactionTime = transactionTime;
    }

    public String getInvestmentId() {
        return investmentId;
    }

    public void setInvestmentId(String investmentId) {
        this.investmentId = investmentId;
    }

    public String getTransactionType() {
        return transactionType;
    }

    public void setTransactionType(String transactionType) {
        this.transactionType = transactionType;
    }

    public BigDecimal getQuantity() {
        return quantity;
    }

    public void setQuantity(BigDecimal quantity) {
        this.quantity = quantity;
    }

    public BigDecimal getPrice() {
        return price;
    }

    public void setPrice(BigDecimal price) {
        this.price = price;
    }

    public BigDecimal getAmount() {
        return amount;
    }

    public void setAmount(BigDecimal amount) {
        this.amount = amount;
    }

    public String getCurrencyCode() {
        return currencyCode;
    }

    public void setCurrencyCode(String currencyCode) {
        this.currencyCode = currencyCode;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public LocalDateTime getProcessDate() {
        return processDate;
    }

    public void setProcessDate(LocalDateTime processDate) {
        this.processDate = processDate;
    }

    public String getProcessUser() {
        return processUser;
    }

    public void setProcessUser(String processUser) {
        this.processUser = processUser;
    }

    public PortfolioMaster getPortfolio() {
        return portfolio;
    }

    public void setPortfolio(PortfolioMaster portfolio) {
        this.portfolio = portfolio;
    }
}
